miner/worker.go

```diff
@@ -101,253 +191,302 @@ type worker struct {
 			// already included in the current mining block. These transactions will
 			// be automatically eliminated.
-			if !self.isRunning() {
-				self.currentMu.Lock()
+			if !w.isRunning() && w.current != nil {
 				txs := make(map[common.Address]types.Transactions)
 				for _, tx := range ev.Txs {
-					acc, _ := types.Sender(self.current.signer, tx)
+					acc, _ := types.Sender(w.current.signer, tx)
 					txs[acc] = append(txs[acc], tx)
 				}
-				txset := types.NewTransactionsByPriceAndNonce(self.current.signer, txs)
-				self.current.commitTransactions(self.mux, txset, self.chain, self.coinbase)
-				self.updateSnapshot()
-				self.currentMu.Unlock()
+				txset := types.NewTransactionsByPriceAndNonce(w.current.signer, txs)
+				w.current.commitTransactions(w.mux, txset, w.chain, w.coinbase)
+				w.updateSnapshot()
 			} else {
 				// If we're mining, but nothing is being processed, wake on new transactions
-				if self.config.Clique != nil && self.config.Clique.Period == 0 {
-					self.commitNewWork()
+				if w.config.Clique != nil && w.config.Clique.Period == 0 {
+					w.commitNewWork()
 				}
 			}
 
 		// System stopped
-		case <-self.txsSub.Err():
+		case <-w.exitCh:
+			return
+		case <-w.txsSub.Err():
 			return
-		case <-self.chainHeadSub.Err():
+		case <-w.chainHeadSub.Err():
 			return
-		case <-self.chainSideSub.Err():
+		case <-w.chainSideSub.Err():
 			return
 		}
 	}
 }
 
-func (self *worker) wait() {
+// seal pushes a sealing task to consensus engine and submits the result.
+func (w *Worker) seal(t *Task, stop <-chan struct{}) {
+	var (
+		err error
+		res *Task
+	)
+
+	if t.block, err = w.engine.Seal(w.chain, t.block, stop); t.block != nil {
+		log.Info("Successfully sealed new block", "number", t.block.Number(), "hash", t.block.Hash(),
+			"elapsed", common.PrettyDuration(time.Since(t.createdAt)))
+		res = t
+	} else {
+		if err != nil {
+			log.Warn("Block sealing failed", "err", err)
+		}
+		res = nil
+	}
+	select {
+	case w.resultCh <- res:
+	case <-w.exitCh:
+	}
+}
+
+// taskLoop is a standalone goroutine to fetch sealing task from the generator and
+// push them to consensus engine.
+func (w *Worker) taskLoop() {
+	var stopCh chan struct{}
+
+	// interrupt aborts the in-flight sealing task.
+	interrupt := func() {
+		if stopCh != nil {
+			close(stopCh)
+			stopCh = nil
+		}
+	}
 	for {
-		for result := range self.recv {
+		select {
+		case task := <-w.taskCh:
+			if w.newTaskHook != nil {
+				w.newTaskHook(task)
+			}
+			interrupt()
+			stopCh = make(chan struct{})
+			go w.seal(task, stopCh)
+		case <-w.exitCh:
+			interrupt()
+			return
+		}
+	}
+}
 
+// resultLoop is a standalone goroutine to handle sealing result submitting
+// and flush relative data to the database.
+func (w *Worker) resultLoop() {
+	for {
+		select {
+		case result := <-w.resultCh:
 			if result == nil {
 				continue
 			}
-			block := result.Block
 
+			block := result.block
```

Comment:
Nitpick, please move this back above the empty line. It's easier on the eye that the first core-block is "initialization" and then an empty line, then comment and continuation.